src/client.rs

```diff
@@ -1286,6 +1286,66 @@ impl<T: Read + Write> Session<T> {
             .and_then(|lines| parse_ids(&lines, &mut self.unsolicited_responses_tx))
     }
 
+    /// The SORT command is a variant of SEARCH with sorting semantics for
+    /// the results. There are two arguments before the searching
+    /// criteria argument: a parenthesized list of sort criteria, and the
+    /// searching charset.
+    ///
+    /// sort            = ["UID" SP] "SORT" SP sort-criteria SP search-criteria
+    /// sort-criteria   = "(" sort-criterion *(SP sort-criterion) ")"
+    /// sort-criterion  = ["REVERSE" SP] sort-key
+    /// sort-key        = "ARRIVAL" / "CC" / "DATE" / "FROM" / "SIZE" /
+    ///                   "SUBJECT" / "TO"
+    /// search-criteria = charset 1*(SP search-key)
+    /// charset         = atom / quoted
+    ///                     ; CHARSET values MUST be registered with IANA
+    /// sort-data       = "SORT" *(SP nz-number)
+    ///
+    /// Below is a selection of common sort keys.  The full list can be found in the
+    /// specification of the [`SORT command`](https://tools.ietf.org/html/rfc5256#section-3).
+    ///
+    ///  - `ARRIVAL`: Internal date and time of the message.  This differs from the ON criteria in SEARCH, which uses just the internal date.
+    ///  - `CC`: IMAP addr-mailbox of the first "cc" address.
+    ///  - `DATE`: Sent date and time, as described in section 2.2.
+    ///  - `FROM`: IMAP addr-mailbox of the first "From" address.
+    ///  - `SIZE`: Size of the message in octets.
+    ///  - `SUBJECT`: Base subject text.
+    ///  - `TO`: IMAP addr-mailbox of the first "To" address.
+    ///
+    ///  - `REVERSE <sort-key>`: Followed by another sort criterion, has the effect of that criterion but in reverse (descending) order.
+    pub fn sort<S: AsRef<str>>(
+        &mut self,
+        criteria: S,
+        charset: S,
+        query: S,
+    ) -> Result<HashSet<Seq>> {
```

Comment:
In fact we can't keep a `HashSet<Seq>` for the `SORT` command, because the order matters.That's why tests fail randomly. I will investigate more.

Reply:
I propose to replace the `HashSet` by a [`LinkedHashSet`](https://crates.io/crates/linked_hash_set), so the insertion order is preserved, what do you think?

Reply:
I'd probably then go with [`IndexSet`](https://docs.rs/indexmap/1/indexmap/set/struct.IndexSet.html) instead, as that crate is much more widely used :)

Another option is to just switch this to a `Vec`, and leave it to the caller whether they actually want a set :sweat_smile: @mordak what do you think?

Reply:
I would be inclined to just use `Vec` - we should not have duplicate IDs in the response anyway (and even if we do it is not catastrophic - the client will just deal with the same ID more than once, which is apparently something the server thinks they should do anyway?). As near as I can tell the original motivation for using `HashSet` was [API convenience](https://github.com/mattnenterprise/rust-imap/pull/89#discussion_r222800445) rather than correctness, so there doesn't seem to be any technical reason for using a `Set`. If we still want things like additional APIs, then `IndexSet` does seem like a more widely used crate.

But since the `SORT` command explicitly cares about the order of responses from the server and ensuring they are passed to the client in that order, then something straightforward like `Vec` is the obvious choice. It is nice to avoid adding additional dependencies for library users if possible, so for me it boils down to if the additional dependency and API uniformity cost is worth the gain. I don't see a really compelling reason to use anything fancy here, so I would just go with `Vec`. :)

Reply:
I proposed this to avoid changing to much the existing code (since `HashSet` and `LinkedHashSet` have more or less the same API), but I totally agree that `Vec` is a much better choice. I will change this whenever I can!

Reply:
Done :heavy_check_mark: